### ISSUE TITLE
Batch Normalization

### ISSUE THREAD

**Issue by skaae** [Lasagne/Lasagne#141]
Moved from https://github.com/benanne/Lasagne/issues/133

**Comment by diogo149** [comment]
Something relevant that I referenced in https://github.com/benanne/Lasagne/issues/136: "If batch normalization mean/variance counts as a param, then SGD update code won't work with it. If it doesn't count as a param, users relying on get_all_params for saving model state will have their code break (because the mean/variance won't be saved)."

**Comment by ebattenberg** [comment]
> Something relevant that I referenced in #136: "If batch normalization mean/variance counts as a param, then SGD update code won't work with it. If it doesn't count as a param, users relying on get_all_params for saving model state will have their code break (because the mean/variance won't be saved)."

I thought the mean/variance were computed for each batch, so they would be intermediate variables and not shared variables/params.  At test time though, they become parameters, but not learned parameters, more like arguments to the function computing the outputs.  

The beta and gamma parameters (from the paper) that transform the normalized outputs *are* learned parameters though.

**Comment by diogo149** [comment]
I did mean at test time, sorry

**Comment by ebattenberg** [comment]
I think that the mean/variance of each unit across the entire training set are pretty specialized parameters in that they have to be computed iteratively layer-by-layer before test time and then communicated to the network at test time.  Because this is different than what is done during training, I'm not sure it makes sense to support this case in the general model serialization code (unless you want to write your own version of model serialization that appends the mean/variance to the serialized model).  Though I haven't dealt with an actual implementation of Batch Normalization yet.

**Comment by benanne** [comment]
That's true, but you also want to get a validation estimate during training so you can track progress. The way they propose to do this in the paper is by using a moving average of the per-batch estimates from training (I'm guessing they mean an exponential moving average which can be implemented more memory-efficiently). But I suppose that's not the only way to do it.

If we do decide to add support code for either of these things (validation estimation, test time estimation), I definitely think it should be separate from our `get_params` infrastructure, which is for learnable parameters only.

I guess this shows that `get_params` is a bit overloaded: it is used both for getting the learnable parameters of a layer, and for getting its state. It seems like those two things do not necessarily overlap completely in all cases. But I don't think this is a strong enough use case to split it up into two separate methods, since they do actually overlap for almost all other use cases.

**Comment by diogo149** [comment]
I agree. I was thinking about this more, and it seems possible to solve this by providing an alternate route for parameter serialization (maybe implementing __getstate__ and __setstate__ to not serialize any theano variables, which would allow serialization of entire models without using get_all_params, and would generally make models easier to read and write).

**Comment by benanne** [comment]
If it's easy to implement those things in the base class and we don't have to require from users to implement them for their own `Layer` subclasses, we can look into it. Otherwise I would consider this a serious increase in complexity for the use case of implementing your own layer, which I definitely think we should avoid.

**Comment by takacsg84** [comment]
Have you guys had any success with this? I gave it a try, and it truly does wonders with the training. It  converges really really quickly, but I`m having problems with the inference part. The variance of batch input means is just too big, so my stored averages don't normalize correctly. If I use (local) batch normalization for inference also, I more or less get back errors of the training, but I don't like the fact that the inference is dependent on the ordering of the test/validation data.

**Comment by ebenolson** [comment]
@takacsg84 can you share your code? I've started working on an implementation of this, but so far only doing normalization (not learning beta/gamma).

**Comment by takacsg84** [comment]
@ebenolson sure!
https://github.com/takacsg84/Lasagne/blob/d5545988e6484d1db4bb54bcfa541ba62e898829/lasagne/layers/bn2.py
While training some additional steps have to be considered. First I get a list of layers that have pre_train and post_train methods:
```python
        # get layers that need pre_train
        pre_train_layers = []
        for l in self.get_all_layers():
            if hasattr(l, 'pre_train'):
                pre_train_layers.append(l)

        # get layers that need post_train
        post_train_layers = []
        for l in self.get_all_layers():
            if hasattr(l, 'post_train'):
                post_train_layers.append(l)
```
then I call these methods before and after every full iteration so I maintain the averages for inference:
```python
            for l in pre_train_layers:
                l.pre_train()

            for Xb, yb in self.batch_iterator_train(X_train, y_train):
                batch_train_loss = self.train_iter_(Xb, yb)
                train_losses.append(batch_train_loss)

            for l in post_train_layers:
                l.post_train()
```

**Comment by ebenolson** [comment]
thanks @takacsg84. When I read the paper before I completely missed the part that said to normalize right before the nonlinearity. What dataset/network are you testing with? I modified your code a bit to do the moving average inside get_output, and it seems to be working ok for me on the MNIST example. I think there may still be a problematic interaction with dropout though.

**Comment by takacsg84** [comment]
@ebenolson 
>When I read the paper before I completely missed the part that said to normalize right before the nonlinearity.

Also don't forget to remove the nonlinearity from the previous layers as well as the bias parameters (as it is redundant).
>I modified your code a bit to do the moving average inside get_output

So you changed the way the means are calculated or just moved it to get_output, so it is only calculated when you actually want inference?
>I think there may still be a problematic interaction with dropout though.

I agree! High dropout could mess up things a bit, but I tested my network w/o dropout, just to be sure.

**Comment by ebenolson** [comment]
>So you changed the way the means are calculated or just moved it to get_output, so it is only calculated when you actually want inference?

I update the moving average each time get_output is called with `deterministic=False`

my changes and test script are in https://github.com/ebenolson/Lasagne/commit/5c0389671d9f9bf2e19a02e4bcfee554659b9e57

**Comment by ebattenberg** [comment]
> Have you guys had any success with this? I gave it a try, and it truly does wonders with the training. It converges really really quickly, but I`m having problems with the inference part. The variance of batch input means is just too big, so my stored averages don't normalize correctly.

At test time, if you're computing the means at every layer independently for each batch, then there could be quite a bit of variation (especially in higher layers) between mini-batches.  I'm not sure what the paper says about this, but I was assuming that in order to get test time estimates of the mean and variance of each unit, you'd want to do something like what I suggested in https://github.com/benanne/Lasagne/issues/141#issuecomment-75175029.  

> I think that the mean/variance of each unit across the entire training set are pretty specialized parameters in that they have to be computed iteratively layer-by-layer before test time and then communicated to the network at test time.

In this approach, you'd compute the mean/variance of the first layer for every mini-batch and then average those.  Then use the average mean/variance across mini-batches to normalize the first layer for every mini-batch in order to compute the second layer.  Then gather statistics for the second layer, and so on.  So it's kind of this iterative thing where the output of a layer requires statistics from the previous layer from every mini-batch.  Make sense?  

Not sure if this is what they communicated in the paper, but I think it's probably the right thing to do at test time.  I'm not really sure about the moving average suggestion for validation during training.  Is that a moving average across validation mini-batches?  If so, how do you normalize the first validation mini-batch?  Or do you keep stats from the previous epoch (in which case the network could have changed quite a bit since)?  Or are they suggesting to interleave validation mini-batches with training mini-batches so that the moving-average is a bit smoother?

This is all a bit confusing since the output of a layer depends on the statistics used to normalize the previous layer which is affected by both the data used to compute the statistics and the changing values of the weights.

**Comment by takacsg84** [comment]
@ebattenberg 
You are totally right about the hierarchical calculation of batch statistics! The paper has no details on this, but it is logical what you suggested. I`ll implement this version also, make some tests, and let you know about the results!

**Comment by ebenolson** [comment]
I think they actually do propose layer-by-layer calculation of the statistics for inference in Algorithm 2. It seems like they use only use examples from the training set to determine the normalization, not validation data. 

The simple moving average calculation I was doing worked ok on the MNIST example, but failed badly on a more complicated test, probably because the normalization is changing more. Replacing it with an exponential moving average improved things, but so far I've gotten the best results just from using the validation mini-batch statistics (non-deterministic).

**Comment by aloisg** [comment]
I'm pretty sure there is a technical limit to this or maybe my idea is wrong, but wouldn't it be possible to evaluate the validation score by processing the validation set as one batch (who would have the size of the whole set) ?
What do i miss ?

**Comment by benanne** [comment]
If your validation or test set is small enough so you won't run out of GPU memory, go for it :) I guess technically some people would consider this cheating because you're calculating statistics on the validation / test set (technically you should calculate the means and variances on the train set and use those).

Unfortunately most validation sets are not small enough for this, in my experience.

**Comment by ebattenberg** [comment]
A couple thoughts:

Regarding Algorithm 2...
> I think they actually do propose layer-by-layer calculation of the statistics for inference in Algorithm 2. It seems like they use only use examples from the training set to determine the normalization, not validation data.

I'm not sure I see this in Algorithm 2.  In the loop starting at line 8,  it looks like they're processing a single hidden node at a time.  I guess if you processed the nodes in order from input to output, you'd effectively be doing this layer-by-layer, but they don't specify this.  


Regarding model serialization...
> I agree. I was thinking about this more, and it seems possible to solve this by providing an alternate route for parameter serialization (maybe implementing __getstate__ and __setstate__ to not serialize any theano variables, which would allow serialization of entire models without using get_all_params, and would generally make models easier to read and write).

An easy way to solve this serialization problem for a model that would only be used for inference (not further training) in the future, would be to compose the normalization with the beta/gamma linear transformation as they suggest at the end of 3.1.  Then you'd just replace beta/gamma (which are Lasagne model parameters) with the composed versions, and the inference model would serialize just fine.  But you'd have to keep the mean/var info around for further training in order to undo the composition.

**Comment by f0k** [comment]
Having finally read the paper myself, I think the batch normalization layer should be structured like this:
* it has `gamma` and `beta` (vectors)
* it also has `mean` and `std` (vectors)
* `get_output_for(input, deterministic=False)` uses `gamma`, `beta` and `T.mean(input)`, `T.std(input)`
* `get_output_for(input, deterministic=True)` uses `gamma`, `beta` and `mean`, `std`
* only `gamma` and `beta` are returned as trainable parameters
* `mean` and `std` should be returned as serializable parameters. There's currently no way to do that.

During training, `mean` and `std` should be updated using a moving average so the network does something useful when computing the validation error (which is usually based on the network outputs with `deterministic=False`). We should provide some helper function to get these updates. Unluckily, this requires the inputs to all batch normalization layers, which means we must compute them along with the network outputs we use for the cost function, which in turn requires #104 to be solved.

After training, `mean` and `std` should be recomputed in the layer-wise fashion [described by Eric](https://github.com/benanne/Lasagne/issues/141#issuecomment-75393349). This falls outside the scope of Lasagne, but again, we could try to provide some help with that.
I think the gains of collapsing `gamma` with `mean` / `beta` with `std` will be neglectible if Theano can be convinced to add / multiply the vectors together before applying them to the input minibatch, so I would avoid that to ensure models can be easily loaded and trained on as if nothing ever happened.

Regarding serialization, we should think again about extending or changing `get_params()` and `get_bias_params()` a bit. With this use case, we have three overlapping classes of parameters:

1. parameters involved in the forward pass (currently `get_params()` according to its docstring)
2. parameters to be updated wrt. the loss function (currently `get_params()` according to its usage in Lasagne)
3. parameters to be updated wrt. the loss function that are not to be included in weight regularization (currently `get_bias_params()`)

From the documentation it is not even entirely clear if `get_params()` returns the first or second category. I'll open up an issue for that (or try to find an existing one, I remember having discussed `get_params()` with Sander before).

**Comment by benanne** [comment]
Thanks for the detailed analysis! I guess we'll have to tackle the API issue first before continuing with this.

**Comment by f0k** [comment]
> During training, `mean` and `std` should be updated using a moving average so the network does something useful when computing the validation error (which is usually based on the network outputs with `deterministic=False`). We should provide some helper function to get these updates. Unluckily, this requires the inputs to all batch normalization layers, which means we must compute them along with the network outputs we use for the cost function, which in turn requires #104 to be solved.

It turns out that Theano has some additional functionality for specifying updates: If a shared variable in a graph has a `default_update` attribute, a corresponding update rule will be created on compilation with `theano.function()` (unless said shared variable is included in the explicit `updates` dictionary or `no_default_updates=True` is given to `theano.function()`).
With some code on the edge between clever and hacky this can be used to incorporate updates to the `std` and `mean` variables in the graph produced with `deterministic=False`, and not include these updates in the graph produced with `deterministic=True`.

> I think the gains of collapsing `gamma` with `mean` / `beta` with `std` will be neglectible if Theano can be convinced to add / multiply the vectors together before applying them to the input minibatch, so I would avoid that to ensure models can be easily loaded and trained on as if nothing ever happened.

Unfortunately, it turns out they use `gamma` and `beta` in the wrong order for them to be collapsed easily. Their formula is: `((input - mean) / std) * gamma + beta`. We can rewrite this to `(input - mean) * (gamma / std) + beta` so at least `gamma` and `std` are merged before applying them to a higher-order tensor, but combining `mean` and `beta` would require some rescaling of `beta`, or changing the learned linear transform to `(x + beta) * gamma`.
So maybe there would be some non-negligible performance gains from manually merging these four parameters into two parameters after training (losing the ability to continue training on them).

---

Anyway, I'm posting my implementation in a gist:
https://gist.github.com/f0k/f1a6bd3c8585c400c190
It doesn't include `mean` and `std` if you only serialize `get_all_param_values()`, and it doesn't include any code for the layer-wise estimation of `mean` and `std` after training, but it can be easily plugged into your architecture by just wrapping every relevant layer in a `batch_norm()` call and then it should do The Right Thing.

**Comment by benanne** [comment]
The `default_update` trick is pretty cool, I didn't know that existed. On the one hand there's a bit of "magic" involved, on the other hand it makes the code look a lot cleaner if this stuff is handled behind the scenes. I'm not 100% sure yet how I feel about using this feature in Lasagne.

I like your proposed implementation, but before adding it to Lasagne I would suggest we deal with #164, so we can incorporate this in it.

**Comment by f0k** [comment]
> I like your proposed implementation

Me too, but I think we shouldn't do this in Lasagne, it's just an implementation for people to use until we have something clean (that depends on #104 and #164).

**Comment by shaih82** [comment]
i never quit realized if the mean is computed 
per pixel which is similar to "mean image" 
or per channel which is RGB mean value.
i just don't know which is better...

**Comment by f0k** [comment]
> i never quit realized if the mean is computed 
> per pixel which is similar to "mean image" 
> or per channel which is RGB mean value.

From the paper, they compute a shared mean and std. dev. per feature map (similar to how you usually have a shared bias per feature map). I.e., if the output of your convolutional layer is `(64, 32, 200, 100)` (32 channels of 200x100, with batch size 64), you would have a mean vector of `(1, 32, 1, 1)` obtained by averaging over all dimensions except for the second. That's what my code does.

**Comment by boptimism** [comment]
I have quite figure out the inference part of that paper. The key, I think, is to compute the moving average of the means and variance. 

>We use the unbiased variance estimate Var[x] = m · EB[σB2 ], where m−1 the expectation is over training mini-batches of size m and σB2 are their sample variances.

But does this process require all the previous epochs? For example, in my training of BN on MNIST, each epoch takes 200 mini-batches. After 2 epochs and 75 mini-batches, I call a stop to the training and start to infer on the test sets. So when I compute the population-wise means and variances, do I use all 475 mini-batches? I got an oscillatory performance of BN. I think the underlying assumptions to use mini-batch samples to reproduce the population stats is that each batch should be independent with each other. This will get violated if once includes all the mini-batches from previous epochs. 

Also, the code doesn't seem efficient to me when it computes the means and variances after parameters gamma, beta, and weights are fixed.
>Ninf ← Ntr // Inference BN network with frozen BN parameters
>Process multiple training mini-batches B, each of size m, and average over them:

If all the previous epochs need to be included, to compute the means and variance for inference is not a pleasant job. And each mini-batch training step can't provide any help for determine the means and variance for inference.

x-axis is **NOT** epochs. Each epoch is reached at interval of 10, e.g. 9, 19, 29, ... indicates an epoch is completed. The black line links all the points corresponding to completed epochs.
![screen shot 2015-03-30 at 8 19 56 pm](https://cloud.githubusercontent.com/assets/4325263/6896142/c190d0a0-d71a-11e4-8fbf-bbce493cbff4.png)

**Comment by boptimism** [comment]
OK, I think once the training is finished, only the training set is used for computing mean/variance that is going to be used by inference step. This way naturally kills the oscillation observed in my previous implementation.
New result:

![screen shot 2015-03-30 at 8 43 56 pm](https://cloud.githubusercontent.com/assets/4325263/6896444/fc947226-d71d-11e4-924b-a0fa20d77f67.png)

**Comment by f0k** [comment]
> The key, I think, is to compute the moving average of the means and variance.

In my initial experiments, I also got good results with just using the (exponential) moving average of means and standard deviations, with `alpha=0.5` [as in my code](https://gist.github.com/f0k/f1a6bd3c8585c400c190).

> So when I compute the population-wise means and variances, do I use all 475 mini-batches?

No, you would freeze gamma and beta (and everything else in the network), then iterate once over your training set (or as much of it as you can afford) in mini-batches of the same size you used before and will use subsequently, computing the mean of all mini-batch means and the mean of all mini-batch standard deviations. You replace the exponential average of means and variances by these true averages and hope that it improves inference. In my case it made things worse, which could have a number of reasons including a bug in my code. I'm postponing further investigations until #164 is done.

**Comment by f0k** [comment]
> This way naturally kills the oscillation observed in my previous implementation.

You shouldn't have oscillations in either case. In your plot, performance drops in every tenth instance. Are you resetting something in between there? The idea of the exponential moving average is to strike a balance between (a) using information of many mini-batches (small alpha) and (b) using information that fit the current network weights (large alpha). It's just meant for monitoring progress, though, the "correct" way is to compute the averages after training with the final set of network weights.

**Comment by boptimism** [comment]
Nope, didn't reset anything in between tenth instances. I am also banging my head all day to try figuring out why the oscillations. It is not clear to me whether previous epochs should be included in the calculation of means and vars that are going to be used in inference stage. I would really appreciate if someone could elaborate on this.

**Comment by f0k** [comment]
> I would really appreciate if someone could elaborate on this.

Didn't I [elaborate enough](https://github.com/benanne/Lasagne/issues/141#issuecomment-87667479)? You want the means and std devs (let's call them "normalization constants") for inference to closely mimic the normalization constants used for training. In training, you use the actual mini-batch mean and std dev, so it's adapted to the current mini-batch and the current network weights. For inference, you cannot use the actual mini-batch mean and std dev, so you use the expected value over your training data, with fixed network weights. For validation during training, it would be expensive to compute the expected value over your training data with fixed network weights, so instead you use a moving average. A moving average takes into account a fixed number of past items (here: mini-batches), not all past items up to now. The more you take into account, the better the estimate because you're using multiple mini-batches, but the worse the estimate because the earlier ones may have had very different network weights (as a few weight updates happened in between). The exponential moving average is a weighted moving average that emphasizes recent mini-batches over older mini-batches, and that can be implemented much memory-efficiently (that's the main reason for using it).

> It is not clear to me whether previous epochs should be included in the calculation of means and vars that are going to be used in inference stage.

Hopefully it's clear now that this question is not really valid. For the inference stage at test time, you don't care about anything that happened during training. For the inference stage at validation time, you include a fixed number of past mini-batches, and you don't care about epochs.

**Comment by boptimism** [comment]
>The more you take into account, the better the estimate because you're using multiple mini-batches, but the worse the estimate because the earlier ones may have had very different network weights (as a few weight updates happened in between).

Wait, please check algorithm 2 in the original paper. The mini-batch's contribution to the means and variance is not added along the training process. Rather, it is when training finished, hence "frozen parameters" in the original paper, all the parameters are used to go back to all the mini-batches. Therefore, for step 10 in algo 2, there is no change for all the parameters, including weights.
I think the workflow is like this:
1. Train BN, obtain weights, gamma, beta
2. Freeze all these parameter; go back to **ALL** the mini-batches used for training; use the frozen parameters to calculate expected means and unbiased variance using moving average/exponential moving average. (*)
3. "Normalize" test data at each layer with newly computed means and variance.

**Comment by benanne** [comment]
No, I'm pretty sure that's wrong. The moving average / exponential moving average is only used for validation DURING training. It makes sense there because the model parameters are constantly changing, so the true value of the means and variances would change over time as well. The moving average just allows you to make a somewhat stable estimate of this without having to do too much computation.

At test time, why would you compute a moving average if you're going to go through all the data anyway? That seems wasteful. Just compute the 'regular' mean and variance. These estimates will be very stable because they are computed on a lot of data (i.e. the entire training set or at least a sizable portion of it). I'm pretty sure that's what the authors of the paper did as well.

**Comment by boptimism** [comment]
@benanne , please check the paper Algorithm 2, step 7-11.
>At test time, why would you compute a moving average if you're going to go through all the data anyway? 

At inference stage, the author did mention:
>Process multiple training mini-batches B, each of size m, and average over them

I was very confused when I first read the paper regarding this step... I guess the logic may be the variance and means from training sets provides good reference for the distribution of the test sets? I also implement an alternative: when test data set size is larger than 1, forget training sets, just do BN on test sets at each layer, meanwhile use the learned weights/gamma/beta to make predictions. This works OK on MNIST.

**Comment by f0k** [comment]
> paper Algorithm 2, step 7-11

This is for post-training, i.e., for computing the test-time parameters. As Sander said, a moving average would be silly here. A moving average is only used for train-time validation. In the paper: "Using moving averages instead, we can track the accuracy of a model as it trains."

> Process multiple training mini-batches B, each of size m, and average over them

This is a plain old average, not a rolling average.

> I guess the logic may be the variance and means from training sets provides good reference for the distribution of the test sets?

That's a basic assumption you do in machine learning. You assume that your training set follows a similar distribution as your test set, in some way or another.

> This works OK on MNIST.

And it has the caveat I've [mentioned above](https://github.com/benanne/Lasagne/issues/141#issuecomment-87685818).

**Comment by boptimism** [comment]
@f0k Thanks for your effort. I agree with what you said about the role of moving average. I guess my real question lies here:
>This is for post-training, i.e., for computing the test-time parameters.

Test-time parameters include population statistics - means and variance from that plain old average. What's the proper way to compute it? It seems to me the authors of that paper use frozen parameters to go through all training mini-batches {B} to obtain the test-time parameters. This part still bothers me. 

1. Correlations among training mini-batches. For example, I used 3 mini-batches to finish training the network, first 2 mini-batches form 1 epoch. Batch 3 and batch 1 are now correlated. Will this have effect on the approximation in the paper: 
> Var[x] = m · EB[σB2 ]/(m-1)

2. I still don't quite understand the caveat part you mentioned. The BN trained network is good at handling data distribution with mean 0 and variance 1. What's inappropriate to normalize the test-set to satisfy the same distribution? The process of transforming x-> xhat is solely dependent on test-set itself. Of course, when there is only 1 test sample, story will be different.

Thanks for your time and patience.

**Comment by f0k** [comment]
> It seems to me the authors of that paper use frozen parameters to go through all training mini-batches {B} to obtain the test-time parameters. This part still bothers me.

In the paper, they write: "Process multiple training mini-batches \mathcal{B}" They leave it open how many mini-batches you use. If you can afford it, use all training mini-batches -- i.e., one full epoch over your training data -- otherwise, use a representative subset. Of course you shouldn't use the same mini-batch multiple times just because it was presented multiple times in training.

> Of course, when there is only 1 test sample, story will be different.

Exactly. What you want is a deterministic mapping from input data to output data. If I have three samples A, B, and C, I want `pred([A]) + pred([B]) + pred([C]) == pred([A,B]) + pred([C]) == pred([A,B,C])`. Note that you don't always have a fixed test set as in MNIST, but you want to build a machine that you can apply to unseen data.

**Comment by boptimism** [comment]
Thanks @f0k . As for the test-time parameters, I think the authors performed a bootstrap on the training mini-batches. That may help figuring out how many mini-batches one needs to consider for inference.

**Comment by cancan101** [comment]
Any updates on this? Or is @f0k's [implementation](https://gist.github.com/f0k/f1a6bd3c8585c400c190) the way to go?

**Comment by f0k** [comment]
My implementation needs to be changed a bit to cater for the new `get_params` API. I already did that, I'll update the gist later. Because of the `default_update` trick for the running mean/std I wouldn't want to include it in Lasagne in this form, but it could become part of the Lasagne Recipes. We can also try to figure out a nicer way of including batch normalization with running mean/std computation for validation, but that can happen after the first release.

**Comment by JackKelly** [comment]
Thanks loads for updating your batch norm implementation, @f0k.

When I first ran it, I got this error:

```
Exception
Traceback (most recent call last):
  File "/homes/dk3810/workspace/python/neuralnilm/scripts/e536.py", line 224, in main
    run_experiment(net, epochs=None)
  File "neuralnilm/experiment.py", line 43, in run_experiment
    net.compile()
  File "neuralnilm/net.py", line 228, in compile
    **self.updates_kwargs)
  File "/homes/dk3810/workspace/python/Lasagne/lasagne/updates.py", line 324, in nesterov_momentum
    updates = sgd(loss_or_grads, params, learning_rate)
  File "/homes/dk3810/workspace/python/Lasagne/lasagne/updates.py", line 134, in sgd
    grads = get_or_compute_grads(loss_or_grads, params)
  File "/homes/dk3810/workspace/python/Lasagne/lasagne/updates.py", line 110, in get_or_compute_grads
    return theano.grad(loss_or_grads, params)
  File "/homes/dk3810/workspace/python/neuralnilm/env/local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 529, in grad
    handle_disconnected(elem)
  File "/homes/dk3810/workspace/python/neuralnilm/env/local/lib/python2.7/site-packages/Theano-0.7.0-py2.7.egg/theano/gradient.py", line 516, in handle_disconnected
    raise DisconnectedInputError(message)
DisconnectedInputError: grad method was asked to compute the gradient 
with respect to a variable that is not part of the computational graph of the cost,
or is used only by a non-differentiable operator: mean
```

I think I've fixed this by passing `trainable=True` to `lasagne.layers.get_all_params`.  

i.e. I now do this:

```python
        all_params = lasagne.layers.get_all_params(output_layer, trainable=True)
```

Is that the correct approach?!

**Comment by f0k** [comment]
> I think I've fixed this by passing trainable=True to lasagne.layers.get_all_params.
> [...]
> Is that the correct approach?!

Yes, that's exactly what's needed. This is actually wrong in the MNIST example, I'll keep that in mind when rewriting it, thanks! The MNIST example still works because all its parameters participate in the forward pass for the training loss, but with batch normalization, some parameters (`mean` and `std`) only participate in the forward pass for the test output and will thus trigger the error you got.

**Comment by JackKelly** [comment]
cool, thanks for the quick reply :)

**Comment by cancan101** [comment]
Does this work for convolutional layers such that the beta and gamma are shared for activations in all locations for a given filter?

**Comment by f0k** [comment]
> Does this work for convolutional layers such that the beta and gamma are shared for activations in all locations for a given filter?

Yes, see the documentation of the `axes` argument: https://gist.github.com/f0k/f1a6bd3c8585c400c190#file-batch_norm-py-L26-L28

**Comment by Ivanopolo** [comment]
f0k, thank you for the code! I couldn't help but notice that after introducing batch norm layers in between my other layers the speed of batch processing dropped almost by a factor of two. Why could that be? Or it's supposed to work that slowly?

**Comment by skaae** [comment]
Will the Batch normalization layer be added to Lasagne? 

@f0k : Can you comment on the difference between the Keras implementation (https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py) and your implementation? 

It seems that Keras does not use the default update trick but saves the updates with self.update? Could something similar be added to lasagne?

**Comment by f0k** [comment]
> Or it's supposed to work that slowly?

It gets quite a bit slower, yes. I don't see a way to speed it up. The idea is that it would still converge more quickly in terms of wall time, but that depends on your data.

> Can you comment on the difference between the Keras implementation and your implementation?

Hmm, they've got two modes: One is using the running batch mean and running batch std deviation (i.e., `axis=0`) to normalize the features, that's what my implementation does for `deterministic=True` to be able to compute the validation error. The other mode is doing something that doesn't have anything to do with the paper, it normalizes the last feature dimension (i.e., `axis=-1`) of each individual training example. They don't have a mode that does what's written in the paper they cite, but the first mode might work similarly well (that would need some experiments).

> It seems that Keras does not use the default update trick but saves the updates with self.update?

Well, the only problem with the default update trick is that it will update the running mean and std deviation if and only if running with `deterministic=False`, assuming that you're doing this for training, and will use the running mean and std deviation instead of the batch-wise mean and std deviation if and only if running with `deterministic=True`. There's currently no way to say "use and update the running mean and std deviation during training" (the first mode in Keras), and there's no way to say "use the batch-wise mean and std deviation, but don't care about updating the running mean and std deviation".

> Could something similar be added to lasagne?

We could add some explicit update mechanism, but that would complicate usage somewhat. It would need to be coupled with `layers.get_output()` or `Layer.get_output_for()`, because that's where we have the input expressions for each layer. So we'd either need to change the API to have `get_output()` return an update dictionary along with the output expression -- that's strongly against our fifth design goal -- or to optionally pass an update dictionary that can be modified by the layers in the `get_output()` call.
I don't see much of an advantage against the current solution of merging the update dictionary with the output expression, though. This, too, could be controlled via some keyword argument to `get_output()` / `get_output_for()`, i.e., we could have `BatchNormLayer` react on some argument like `update_norms` that overrides the assumption based on `deterministic=True` or `deterministic=False`. And it relieves the user from having to obtain a second update dictionary to keep and pass on to `theano.function`.

So unless we find a use case that doesn't work well with default updates, I guess we could leave the API as it is and merge the batch normalization layer as it is. I'll make a PR some time soon to move the discussion there!

**Comment by f0k** [comment]
> I don't see a way to speed it up.

Well, there'd be one... in the backward pass, the gradient now also flows back through the `mean()` and `std()` operations, but one could try treating the batch-wise mean and standard deviation as constants for the backward pass (basically, that's what Keras does when using a shared variable instead of the direct expressions). In the original paper, the authors do propagate back through the batch-wise mean and standard deviation, though (p. 4, after "We use chain rule" [sic]).

**Comment by benanne** [comment]
Isn't one of the points they make that it's important to also backpropagate through the normalisation operation? Or am I confusing this with another paper? :)

I know of a good way to speed it up though: write a custom CUDA kernel and wrap it with PyCUDA :p

**Comment by Ivanopolo** [comment]
Just interesting, what kind of a speed up would it give, in your opinion?

понедельник, 10 августа 2015 г. пользователь Sander Dieleman написал:

**Comment by f0k** [comment]
> Isn't one of the points they make that it's important to also backpropagate through the normalisation operation? Or am I confusing this with another paper?

That would be interesting, but I don't remember reading that. If anybody wants to re-read the paper to check, let us know! I won't experiment with it any time soon.

> Just interesting, what kind of a speed up would it give, in your opinion?

I don't think it would be easy to write a kernel that's a lot faster than what Theano does by default. But not propagating through the normalisation operation would remove most of the backpropagation costs. It's a bit more difficult to implement (you'd need to wrap the operation in an Op and override the gradient definition, this could be done via [`OpFromGraph`](http://deeplearning.net/software/theano/library/compile/opfromgraph.html) and setting its `grad()` member to a new function). And we don't know if that would defeat the purpose, you'd need to check if Sander's memory of the paper is better than mine in this respect.

**Comment by benanne** [comment]
The importance of backpropagating through the normalization step + parameters is discussed in Section 2, paragraph 2 and onwards of the Batch Normalization paper. It goes on for about half a page, so the authors seemed to think it's pretty important to do this.

I believe a CUDA kernel could actually speed things up quite a bit, because batchnorm involves a lot of elementwise and reduction operations that could be fused into a single kernel to avoid spurious memory allocations. I'm sure Theano already does its best to avoid some of those, but I doubt the optimizations are smart enough to get all of them. The best way to find out would be to study the compiled graphs, I guess.

**Comment by f0k** [comment]
> The importance of backpropagating through the normalization step + parameters is discussed in Section 2, paragraph 2 and onwards of the Batch Normalization paper.

Kudos to your memory! And the argument is quite compelling, too. So the implementation in Keras shouldn't actually work (using a shared variable to buffer the normalization parameters prevents the gradient from flowing back through the normalization operations).

**Comment by skaae** [comment]
Do you mean that the gradient cannot flow through this line:

https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L49

Becasue the running means are shared variables? I agree that they do not seem to implement whats in the paper.

Your implementation directly use the minibatch mean and std during training and does not have that problem?

I might be lost in the implementation.

**Comment by f0k** [comment]
> Do you mean that [...]

Yes!

> Your implementation directly use the minibatch mean and std during training and does not have that problem?

Yes!

> I might be lost in the implementation.

You seem fine :)

**Comment by nouiz** [comment]
About Theano speed, I checked someone else graph of this some time ago. We
are missing at least one elemwise optimization. We only merge elemwise with
the condition that the fused elemwise have only 1 output. That layer would
need elemwise with multiple outputs. I started work on this. this PR:

https://github.com/Theano/Theano/pull/3263

Contain fix/update to allow someone to manually create an Elemwise with
multiple output and it should work on CPU and GPU with inplace
optimization. You could try this easily by manually making the elemise like
in this test file:

https://github.com/Theano/Theano/pull/3263/files#diff-d7b215421f4479ae79ef0e6408ba27f3

I started work a few months ago to implement the optimization to make this
automatic, but it won't be ready in the next few weeks.

Hopefully, the PR can be merged next week to allow the manual case.

**Comment by f0k** [comment]
Looks promising, but it's not too easy to use it manually for this case -- at least we'd need to define the gradient by hand then. Still good to hear there's some more potential!

**Comment by nouiz** [comment]
I forgot that the grad wasn't defined. We could implement Composite.grad.
It should be very similar to OpFromGraph.grad.

I made a ticket, not sure when it will be done:
https://github.com/Theano/Theano/issues/3272

**Comment by f0k** [comment]
> I forgot that the grad wasn't defined. We could implement Composite.grad.

But the gradient would need to become a multiple-output elemwise Op again, can Composite.grad do this at all? Maybe it's not worth doing a general implementation of Composite.grad, all this will become unnecessary once Theano introduces the multi-output elemwise Op itself via an optimization.

**Comment by nouiz** [comment]
Composite support multiple output for a long time. It was some
optimization(causing graph) and the GpuElemwise(also causing crash) that
didn't supported that.

In fact, the more I think about this, the more I think it is easy to do,
but I have other stuff to do today and won't be available until 2-3 weeks.
A quick implementation copied from OpFromGraph:

    def grad(self, inputs, output_grads):
        gs = theano.gradient.grad(cost=None,
                                  known_grads=dict(izip(self.outputs,
                                                        output_grads)),
                                  wrt=self.inputs,
                                  disconnected_inputs='ignore')
        new_op = Composite(self.inputs + output_grads, gs)
        return new_op(*(inputs + output_grads))



If you wan to try it. The only inconvenince is that it will except all
inputs of the forward and all the grad signals to be needed for the grad op.

I think this could be worked around with something like this:

    def grad(self, inputs, output_grads):
        gs = theano.gradient.grad(cost=None,
                                  known_grads=dict(izip(self.outputs,
                                                        output_grads)),
                                  wrt=self.inputs,
                                  disconnected_inputs='ignore')
        used_inputs = theano.gof.graph.inputs(gs)
        new_inputs_inner = []
        new_inputs_outer = []
        for in_i, out_i in zip(inputs, self.inputs):# + output_grads:
           if in_i in used_inputs:
                new_inputs_inner.append(in_i)
                new_inputs_outer.append(out_i)
        for o_grad output_grads:
           if o_grad in used_inputs:
                new_inputs_inner.append(o_grad)
                new_inputs_outer.append(o_grad)

        new_op = Composite(new_inputs_inner, gs)
        return new_op(*new_inputs_outer)

It also need the connection pattern, this work with the first version of
grad above, but not the second (need update for the inputs I think.

    def connection_pattern(self, node):
        """
        Return connection pattern of subfgraph defined by inputs and outputs
        """
        return io_connection_pattern(self.inputs, self.outputs)

**Comment by blackyang** [comment]
Hi, @f0k , when using your [implementation](https://gist.github.com/f0k/f1a6bd3c8585c400c190), I always got errors when using momentum, adadelta (not sure for others) updates **except** rmsprop. The errors are like below:

Error when tring to find the memory information on the GPU: an illegal memory access was encountered
Error freeing device pointer 0x1322760000 (an illegal memory access was encountered). Driver report 0 bytes free and 0 bytes total
CudaNdarray_uninit: error freeing self->devdata. (self=0x7fdb363aedf0, self->devata=0x1322760000)
Error when trying to find the memory information on the GPU: an illegal memory access was encountered
RuntimeError: Failed to allocate output
Apply node that caused the error: GpuCAReduce{add}{1,0,1,1}(GpuElemwise{Composite{((i0 * i1) + (i0 * i1 * sgn(i2)))}}[(0, 1)].0)
Toposort index: 520

Any ideas to solve this problem? Thanks!

**Comment by benanne** [comment]
Well, this looks interesting, let's keep an eye on it :)
https://github.com/Theano/Theano/pull/3410

**Comment by lc82111** [comment]
Hi, @f0k 
After applying your BN code https://gist.github.com/f0k/f1a6bd3c8585c400c190#file-batch_norm-py-L93 on top of my mnist mlp network, I got NaN in the training loop. Here is my code (use nolearn framewok):
```
def build_mlp(input_var=None):
    l = InputLayer(shape=(None, 1, 28, 28), input_var=input_var)
    l = ReshapeLayer(l, ([0], 28*28))
    l = DenseLayer(l, num_units=800,b=None, nonlinearity=None,)
    l = BatchNormLayer(l, nonlinearity=tanh)
    l_out = DenseLayer(l, num_units=10,nonlinearity=softmax)
    return l_out

l_out = build_mlp()

net_mnist = NeuralNet(
    layers=l_out,
    max_epochs=200,
    update=adam,
    #update = nesterov_momentum,
    update_learning_rate=theano.shared(float32(0.0001)),
    #update_momentum=theano.shared(float32(0.9)),
    on_epoch_finished=[
        #AdjustVariable('update_learning_rate', start=0.0001, stop=0.00001),
        #AdjustVariable('update_momentum', start=0.9, stop=0.999),
        EarlyStopping(patience=20),
        ],
    batch_iterator_train=BatchIterator(batch_size=512),
    batch_iterator_test=BatchIterator(batch_size=512),
    objective=regularization_objective,
    objective_lambda2=0,

    train_split=TrainSplit(eval_size=0.20),
    verbose=1,
)
X, y = load_mnist_dataset()
net_mnist.fit(X,y)
```
Below is NanGuardMode output:
```
Apply node that caused the error: Elemwise{true_div}(Elemwise{mul}.0, Elemwise{Abs}.0)
Toposort index: 235
Inputs types: [TensorType(float32, matrix), TensorType(float32, matrix)]
Inputs shapes: [(1, 800), (1, 800)]
Inputs strides: [(3200, 4), (3200, 4)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[GpuFromHost(Elemwise{true_div}.0)]]

Debugprint of the apply node: 
Elemwise{true_div} [@A] <TensorType(float32, matrix)> ''   
 |Elemwise{mul} [@B] <TensorType(float32, matrix)> ''   
 | |Elemwise{second} [@C] <TensorType(float32, matrix)> ''   
 | | |Elemwise{abs_,no_inplace} [@D] <TensorType(float32, matrix)> ''   
 | | | |HostFromGpu [@E] <TensorType(float32, matrix)> ''   
 | | |   |beta [@F] <CudaNdarrayType(float32, matrix)>
 | | |DimShuffle{x,x} [@G] <TensorType(float32, (True, True))> ''   
 | |   |Elemwise{mul} [@H] <TensorType(float32, scalar)> ''   
 | |     |Elemwise{second,no_inplace} [@I] <TensorType(float32, scalar)> ''   
 | |     | |Elemwise{add,no_inplace} [@J] <TensorType(float32, scalar)> ''   
 | |     | | |Elemwise{true_div,no_inplace} [@K] <TensorType(float32, scalar)> 'mean'   
 | |     | | |Elemwise{add,no_inplace} [@L] <TensorType(float32, scalar)> ''   
 | |     | |   |Elemwise{mul,no_inplace} [@M] <TensorType(float32, scalar)> ''   
 | |     | |   | |TensorConstant{0.0} [@N] <TensorType(float32, scalar)>
 | |     | |   | |Elemwise{add,no_inplace} [@O] <TensorType(float32, scalar)> ''   
 | |     | |   |   |Elemwise{add,no_inplace} [@P] <TensorType(float32, scalar)> ''   
 | |     | |   |   | |Sum{acc_dtype=float64} [@Q] <TensorType(float32, scalar)> ''   
 | |     | |   |   | | |Elemwise{abs_,no_inplace} [@R] <TensorType(float32, matrix)> ''   
 | |     | |   |   | |   |HostFromGpu [@S] <TensorType(float32, matrix)> ''   
 | |     | |   |   | |     |W [@T] <CudaNdarrayType(float32, matrix)>
 | |     | |   |   | |Sum{acc_dtype=float64} [@U] <TensorType(float32, scalar)> ''   
 | |     | |   |   |   |Elemwise{abs_,no_inplace} [@D] <TensorType(float32, matrix)> ''   
 | |     | |   |   |Sum{acc_dtype=float64} [@V] <TensorType(float32, scalar)> ''   
 | |     | |   |     |Elemwise{abs_,no_inplace} [@W] <TensorType(float32, matrix)> ''   
 | |     | |   |       |HostFromGpu [@X] <TensorType(float32, matrix)> ''   
 | |     | |   |         |W [@Y] <CudaNdarrayType(float32, matrix)>
 | |     | |   |Elemwise{mul,no_inplace} [@Z] <TensorType(float32, scalar)> ''   
 | |     | |     |TensorConstant{0} [@BA] <TensorType(int8, scalar)>
 | |     | |     |Elemwise{add,no_inplace} [@BB] <TensorType(float32, scalar)> ''   
 | |     | |       |Elemwise{add,no_inplace} [@BC] <TensorType(float32, scalar)> ''   
 | |     | |       | |Sum{acc_dtype=float64} [@BD] <TensorType(float32, scalar)> ''   
 | |     | |       | | |Elemwise{pow,no_inplace} [@BE] <TensorType(float32, matrix)> ''   
 | |     | |       | |   |HostFromGpu [@BF] <TensorType(float32, matrix)> ''   
 | |     | |       | |   | |W [@T] <CudaNdarrayType(float32, matrix)>
 | |     | |       | |   |DimShuffle{x,x} [@BG] <TensorType(int8, (True, True))> ''   
 | |     | |       | |     |TensorConstant{2} [@BH] <TensorType(int8, scalar)>
 | |     | |       | |Sum{acc_dtype=float64} [@BI] <TensorType(float32, scalar)> ''   
 | |     | |       |   |Elemwise{pow,no_inplace} [@BJ] <TensorType(float32, matrix)> ''   
 | |     | |       |     |HostFromGpu [@BK] <TensorType(float32, matrix)> ''   
 | |     | |       |     | |beta [@F] <CudaNdarrayType(float32, matrix)>
 | |     | |       |     |DimShuffle{x,x} [@BL] <TensorType(int8, (True, True))> ''   
 | |     | |       |       |TensorConstant{2} [@BH] <TensorType(int8, scalar)>
 | |     | |       |Sum{acc_dtype=float64} [@BM] <TensorType(float32, scalar)> ''   
 | |     | |         |Elemwise{pow,no_inplace} [@BN] <TensorType(float32, matrix)> ''   
 | |     | |           |HostFromGpu [@BO] <TensorType(float32, matrix)> ''   
 | |     | |           | |W [@Y] <CudaNdarrayType(float32, matrix)>
 | |     | |           |DimShuffle{x,x} [@BP] <TensorType(int8, (True, True))> ''   
 | |     | |             |TensorConstant{2} [@BH] <TensorType(int8, scalar)>
 | |     | |TensorConstant{1.0} [@BQ] <TensorType(float32, scalar)>
 | |     |TensorConstant{0.0} [@N] <TensorType(float32, scalar)>
 | |HostFromGpu [@E] <TensorType(float32, matrix)> ''   
 |Elemwise{Abs} [@BR] <TensorType(float32, matrix)> ''   
   |HostFromGpu [@E] <TensorType(float32, matrix)> '' 
```

**Comment by f0k** [comment]
> It seems the div op generats NaN in gradient backward propagation?

You may try replacing `std = input.std(self.axes, keepdims=True)` with `std = (input.var(self.axes, keepdims=True) + self.eps).sqrt()`, and remove the `std += self.epsilon` further below. It should have been like that in the first place. But I can't tell if this is related to your error.

Regarding your NaNGuardMode output: There are a lot of `HostFromGpu` in the graph that copy weight matrices from GPU to CPU, then perform some operation on CPU in double precision, and copy things back (looks like an L2 regularizer, as it squares and sums the weight matrices). This may not have anything to do with your problem, but you should have a look into that, it will slow things down a lot.

> Using mnist example code shipped with lasagne, every thing seems right. I will check nolearn.

Try to find out what's different. The MNIST example in Lasagne doesn't have L2 regularization, for example.

**Comment by hafssol** [comment]
I Implemented moving average scheme for theano's BN usage.
It worked pretty fine for my network.  (4 CONV + 1 MLP + 1 LSD) 
I applied BN to CONV only.
This way infers only last mean/std, it will save the memory. 

   for every mini-batch update shared variable mean, std
      *mean = (1-e) * mean + e * map.mean()
      *std = (1-e) * std + e * map.std()
   end
I used e = 0.01 (I think last of training, e should be near 0)

**Comment by dangweili** [comment]
Will the BN layer works correctly when there are multiple channels, such as simase and triplet ?
In these multiple channel networks, the value in BN layer should be the same (shared) in all channels. However, I think the current version can not work as that. Any one has some suggestions ?

**Comment by f0k** [comment]
> Any one has some suggestions ?

The best (most performant) way to implement a siamese network is to make your batches twice as large (concatenating the examples that'd feed into the left and right leg) and splitting or reshaping them when they reach the head (with a SliceLayer to obtain the left and right part, or a ReshapeLayer to rearrange the tensors such that the left and right parts are concatenated). If you do this, the BN layer will normalize values using all the inputs simultaneously. Otherwise, the only way is to write a custom BN layer that accepts and normalizes multiple inputs.
